pkg/aws/sts/cache.go

```diff
@@ -41,19 +42,19 @@ type RoleCredentials struct {
 }
 
 const (
-	DefaultPurgeInterval             = 1 * time.Minute
-	DefaultCredentialsValidityPeriod = 15 * time.Minute
-	DefaultCacheTTL                  = 10 * time.Minute
+	DefaultPurgeInterval = 1 * time.Minute
+	DefaultCacheTTL      = 10 * time.Minute
```

Comment:
It probably just needs to be something like `CacheTTL - (2 * time.Minute)`

Reply:
Good point about the durations being controlled by AWS. We should probably rename most of those to be something like `requestedTTL` to indicate it's what we're asking, rather than the actual duration of the token.

The future is there to prevent us requesting credentials for the same role multiple times- since access to the cache is under a Mutex if multiple clients request credentials for the same role they'll be given references to the same `Future`, so they can all block on the same request for credentials and complete when the credentials are returned. It's somewhat of an optimisation but means that the number of AWS calls reduces significantly.

I'm not sure if there's any guarantees from AWS about the duration- if the operation can _only_ succeed when credentials are returned with the requested TTL then I'd leave things mostly as-is. However, if STS can disregard what you request and override the duration (so the expiry on the returned response is the only thing that matters) then we'll probably need to re-jig things a bit.

What do you think?

Reply:
Ok, that makes sense. So we could simply recache the entry with the new expiry time since the cache backend will overwrite the cache copy with the new entry and TTL.

So as summary my plan is:
- We know that STS can return different lifetime as what is requested
- We will first cache the futures version of the request using requestedTTL as guideline (essentially we should have a shorter TTL for unresolved entries)
- Once the request is resolved, we will recache the received entry with the actual TTL (minus grace period)

This should allow the current logic to work as intended. In worst case scenario, this will cause the same entry to be rewritten multiple times to cache, which should have minimal performance impact in the big picture.

Does that sound reasonable?

Reply:
I don't think it's safe to do the recache- the mutex will only protect around the manipulation of the item with that key- to recache would mean different operations against the cache with no mutex around them. We could add our own synchronisation in but it'd be nicer to lean on the underlying data structures managing (i.e. the cache and the future).

Let me do some reading and see if the AWS API makes any statements around the durations, not sure whether returning credentials with a different expiry is possible/erroneous etc.

Reply:
I've not found anything obvious. My assumption is that the duration returned is used to track the expiry of the issued credentials rather than an assertion that AWS would return credentials with a different duration than you request. 

Their documentation does specify a requirement that the duration is between 15 minutes and 1 hour. I would expect that if you request something longer (or shorter) you'd receive an error, rather than the response containing credentials with a different validity duration. Before doing more I'd want to test this (should be easy enough at the console using the AWS cli- if I get a chance later I'll give it a go).

Potentially we could record a warning/something if the issued credentials have a duration that's significantly different than what's requested.

Reply:
Recache should be safe, since we would be recaching the same entry. According to https://godoc.org/github.com/patrickmn/go-cache#Cache.Set the cache will simply be overwritten and looking at the code itself, it does locking for the modification.

So in essence in cache.go we would be doing something in the lines of:
```go
f := future.New(issue)
c.cache.Set(role, f, DefaultCacheTTL)

val, err := f.Get(ctx)
if err != nil {
    c.cache.Delete(role)
    return nil, err
}

// parse returned expiration and calculate TTL duration
c.cache.Set(role, f, NewCacheTTL)
[...]
```

This should allow everything else to work normally. Limiting the range would also be possible, but considering that there is an expiration in the response, it makes sense to use that expiration as the cache TTL.

My concern here is mostly that since AWS is the source of the expiration, they might return an unexpected value (aka. shorter expiration). If that happens, the token will be invalid until we reach the default TTL. In this sense the 10 minute default value is a good safeguard. The session can never be less than 10 minutes so there is no situation when it might expire while still being cached. I haven't seen this happen, but seeing how they allow this behavior in `GetSessionToken` (for example) where you are allowed to define a value up to 12h and they will simply return 1h token in some cases, I wouldn't put be surprised if the same logic applied in here as well.

Reply:
We just had a discussion here about the caching and during that discussion I realised that it's actually better if we keep the cache TTL 10 minutes. That will spread out the token renewals for the pods themselves.

Let's assume that we renew the token 2 minutes before the token expiration. this means that for 13 minutes all botocore based applications will refresh tokens constantly prior to refresh. Even if that isn't the case, this will cause a storm of requests that all coincide with the token expiration within kiam, effectively causing spikes in load for kiam. Refreshing cached tokens every 10 minutes will still reduce the load on AWS API quite significantly, while avoiding request storms.

Reply:
I don't believe you can safely set the cache, retrieve and then set again- the mutex applies within those functions and not around them so we'd need other synchronisation to happen in your example code above.

Is your suggestion then that we'd keep all cache expiries/refreshes to 10 minutes, but that credentials would be valid for longer- so clients would refresh them less frequently against kiam but the kiam server would be interacting with AWS every 10 minutes?

Reply:
The cache does support a [Replace operation](https://github.com/patrickmn/go-cache/blob/master/cache.go#L106) (and could probably be updated to support modifying an item's expiry) that has the correct synchronisation.

To summarise, it feels like the options are:

1. Change the cache to allow expiries to be updated/changed (either with a change to the lib, or using the `Replace` function. Cache expiry would be set to `Expiry - (time.Minute * jitter)` to ensure its refreshed ahead of client's requesting credentials.
2. Allow sessions to have configurable durations but maintain a fixed refresh interval (via the cache expiration) of 10 minutes.

Perhaps the safest thing for now is option 2 (which is pretty much what you originally contributed via this PR). It means that `--session-duration` _can't_ be used to reduce load on the AWS API but would allow clients to need credentials less frequently. 1 is probably better but I'd want to be sure that it was safe under load.

Reply:
I'm currently in favour of option 2. My original concern wasn't to reduce load on AWS api, but rather on kiam (and the system doing API calls through kiam). The problem we encountered was to do with latency. Every request received by the botocore backed python app caused 4 HTTP requests to be triggered to request instance credentials. Increasing the session duration caused the caching within botocore to hold the keys in memory and reduced the overhead from credential fetching.

I'll update the PR with the naming changes only for now.

Reply:
Also a related issue was with credential fetching where kiam was unable to return credentials to the app, but I'll try and figure out the root cause for this later on.